### DISCUSSION TITLE
strtuctured / nui chat sidebar

### DISCUSSION THREAD

**Discussion by dlants** [olimorris/codecompanion.nvim#495] [Ideas]
As a basic example, when running in agent mode, the LLM may choose to use several tools simultaneously. Immediately dropping into the first diff, without being able to see the context of the entire change that the agent is trying to make, is really disorienting.

I think a UI would ideally:
- create UI elements for each proposed tool use
- for edits, allow you to view the proposed changes in the chat buffer, and also in the main buffer as a diff. Allow you to re-open the proposed diff as you go (kind of like windsurf). Allow you to apply the diff from the chat buffer.
- for tools that don't require permission, execute them in parallel and in the background, but display their status in the chat ui (like reading a file, or running RAG or whatever else).
- allow you to navigate to the tools requiring permission in the chat buffer and approve/reject them using keyboard commands

I think this also has a likelihood of helping with some other things like:
- not interfering with the chat buffer while it's streaming a response
- keeping the input & the scrollback separate so you can always get back to typing the next message quickly, and see the scrollback while composing your message
- having a space to see the shared context and tools, and revoke their permissions.

I think going in the direction of avante, which seems to use nui, is probably a good option to consider: https://github.com/yetone/avante.nvim/blob/main/lua/avante/sidebar.lua

**Comment by dlants** [comment]
Ah apologies, just noticed https://github.com/olimorris/codecompanion.nvim/discussions/33

**Comment by olimorris** [comment]
I would be really keen to see what we could do with nui and how we could improve the overall workflow for a user.

My philosophy for CodeCompanion will always be that it enables simple interaction with an LLM via a native buffer. It was ChatGPT.nvim forcing me to use floating windows and a million keymaps that led to me building CodeCompanion (courtesy of a helpful foundation from SteveArc).

My views on UI libraries like nui is that they tread a fine line between being vim-like and a GUI editor. That being said, I think most of your suggestions sound fantastic and for a lot of users would be very welcome. So the option of having a "classic" and "modern" UI could be very cool indeed.

**Comment by bassamsdata** [comment]
I have some thoughts on this. I believe a chat mechanism should prioritize ease of use. Unlike `chatgpt.nvim` (or any multi-floating-window approach), which requires users to memorize **keymaps** for navigating between windows—something not very practical—the experience should feel more intuitive.  

An exception I’ve seen is `avante.nvim`, which uses NUI with two windows and allows simple **Tab** navigation for interaction. However, I personally prefer interacting with chat in a normal buffer. The downside of a normal buffer approach is that the user can’t start typing until the LLM finishes streaming.  

In my opinion, the best LLM chat implementation in any editor is `Zed`. It feels similar to `CodeCompanion`, offering the familiarity of Neovim buffers that can accept any Vim motions, **a big advantage**. Zed’s chat buffer advantage is allowing users to write while the LLM continues streaming.  

I think `CodeCompanion` could adopt a similar approach by enabling users to write in a small bottom floating window with hidden borders. It should have most features that making it feel integrated into the buffer. Once the LLM finishes, it could automatically paste the user's text into the original chat buffer.

the below gif in zed to show how zed works for people didn't try zed chat.
![Screen Recording 2024-12-03 at 11 58 36 AM](https://github.com/user-attachments/assets/2244121a-66c6-4516-be30-19f3b3ee8998)

**Reply by bassamsdata** [reply]
The additional features suggested in the original post seem pretty cool to include.
